src/layers/icmp.rs

```diff
@@ -0,0 +1,107 @@
+//! ICMP Datagram
+
+use std::convert::TryInto;
+
+use serde::Serialize;
+
+use crate::errors::Error;
+use crate::layers::ipv4;
+use crate::types::IPv4Address;
+use crate::Layer;
+
+/// IANA Assigned protocol number for ICMP
+pub const IPPROTO_ICMP: u8 = 1_u8;
+
+// Register ICMP with Protocol Handler in IPv4
+pub(crate) fn register_defaults() -> Result<(), Error> {
+    ipv4::register_protocol(IPPROTO_ICMP, ICMP::creator)
+}
+
+#[derive(Default, Debug, Serialize)]
+#[serde(untagged)]
+pub enum IcmpType {
+    #[default]
+    Unused,
```

Comment:
Also, `Unsupported` variant can have `Vec<u8>` as additional data, so something like `Unsupported(Vec<u8>)`. The reason being, during parsing if we encounter an `Unsupported` variant, remaining bytes won't be accounted for and would be given to `Packet`'s remaining data, but they actually belong to `ICMP` layer.

Reply:
We can add `Unsupported` as `u32` as it holds the unsupported part of the header. The part which belongs to the ICMP packet data can be handled by the IcmpData enum - which I am working on currently.

Reply:
Well `u32` is semantically not correct. It would be a different number based on endianness of the machine, so if a particular ICMP type is not supported, just dumping everything into a `Vec<u8>` is a good idea.

Reply:
Yes, sir, I had some confusion. I have made the appropriate changes.